port/cpl_vsi_mem.cpp

```diff
@@ -70,6 +72,38 @@
 
 //! @cond Doxygen_Suppress
 
+// szHIDDEN_DIRNAME is for files created by VSIMemGenerateUnlistedFilename(pszFilename).
+// Such files are of the form "/vsimem/.#!HIDDEN!#./{counter}/{pszFilename}"
+//
+// The high-level design constraint is that "/vsimem/.#!HIDDEN!#." acts as a
+// "side" hiearchy, but still unders the "/vsimem/" namespace, so that code
+// acting specifically on /vsimem/ files. The structure of the returned filename
+// is also such that those files form independent hierarchies, i.e. the tree
+// generated by a VSIMemGenerateUnlistedFilename() is "invisible" from the one
+// returned by another call to it.
+//
+// As a consequence:
+// - we don't want ".#!HIDDEN!#." to be listed in VSIReadDir("/vsimem/")
+// - we don't want content under ""/vsimem/.#!HIDDEN!#" to be deleted by
+//   VSIRmdirRecursive("/vsimem/")
+// - we don't want the creation of a file (or directory) called
```

Comment:
I'm having trouble understanding this. I would have expected 

- directories starting with `.#!HIDDEN!#.'` are not listed
- the filename returned by the new function is in a regular directory with a name like like `mktemp -u -d -p '/vsimem/.#!HIDDEN!#.'`. 

It sounds more complex than this. I would need to check it out and play around to understand it.

Reply:
Imagine:
- First call with VSIMemGenerateHiddenFilename("foo.gml") will return: "/vsimem/.!HIDDEN!./1/foo.gml"
- Second call with VSIMemGenerateHiddenFilename("bar.shp") will return: "/vsimem/.!HIDDEN!./2/bar.shp"
- Third call with VSIMemGenerateHiddenFilename("my_dir") will return: "/vsimem/.!HIDDEN!./3/my_dir", and user calls VSIMkdir("/vsimem/.!HIDDEN!./3/my_dir"), and craetes files "one" and "two" within it

Then
- VSIReadDir("/vsimem/") does not mention ".!HIDDEN!"
- VSIReadDir("/vsimem/.!HIDDEN!.") returns "1", "2" and "3", but this is an implementation detail, just for the purpose of our unit tests as code is not suppose to access the parent directories of the return path
- VSIReadDir("/vsimem/.!HIDDEN!./3/my_dir") returns "one" and "two"

Reply:
> Why is "2" not just a regular directory with a unique name?

"Explicit" directories, like regular files, are listed in the oFileList member variable. If they are listed there, someone must clean up them at some point using Unlink() or Rmdir().
For a file returned by VSIMemGenerateHiddenFilename("foo.tif"), so something like "/vsimem/.!HIDDEN!./123/foo.tif",  we only want the returned filename to be something explicit, because that's the only resource the user must take care of deleting, but not its parent directories "/vsimem/.!HIDDEN!./123" or ""/vsimem/.!HIDDEN!."
Hence VSIReadDir("/vsimem/.!HIDDEN!."), which is normally just for debugging purposes, has special processing to synthetize the "{counter}" directories from the filenames like "/vsimem/.!HIDDEN!./123/foo.tif", but without needing an explicit "/vsimem/.!HIDDEN!./123" entry in oFileList
We could possibly have VSIStat() work on "/vsimem/.!HIDDEN!./123" but that doesn't appear to me as being strictly needed for now.

Reply:
I understand how treating these directories as a special case avoids the need to clean them up later. And I guess forcing every temporary file into a separate "directory" avoids potential collisions from sidecar files. It seems like a complex solution relative to following an existing API like `mktemp`, but not something I feel terribly strongly about.